Project review

perisicMarko opened the pull request:
U template fajlu su bloom i fbuff w pp lekcije iz iste grupe, to sam video tek kada sam zavrsio projekat. Na sajtu stoji da su to lekcije iz razlicite grupe. Sta je predvidjeno za ovu godinu, to iz template-a i ili informacije sa sajta?

perisicMarko committed:
Build plain app
Set app to close on esc

perisicMarko committed:
rename the window title
add skybox

perisicMarko committed:
change skybox
add moon model

perisicMarko committed:
add wasd(keyboard) camera navigation
make observer for mouse change
add depth testing
set cursor hidden

perisicMarko committed:
add space_station model
change window title to Space station
set first mouse flick skipped

perisicMarko committed:
add space_craft model
change window size
adjust size and position of all added models

perisicMarko committed:
add pointer light from moon source to ship and station

perisicMarko committed:
update PROJECT-DESCRIPTION-TEMPLATE.md
add TODO list

perisicMarko committed:
add gui with customizable elements
update TODO.md and PROJECT-DESCRIPTION-TEMPLATE.md

perisicMarko committed:
Svaki fragment position vertex arguemnt mi je nula, u bilo kom modelu
stavio sam da se modeli boje u crveno ako im je fpos = 0

perisicMarko committed:
post pull request code refactoring
-cursor stats from Camera.hpp to PlatformController.hpp
-minor code cleansing
-private class fields renamed to m_ + field_name

perisicMarko committed:
add spotlight to scene
add spotlight components sliders into gui

perisicMarko committed:
correct addition of light components in shaders

perisicMarko committed:
adjust models positions
add moon event on E

perisicMarko committed:
add spacecraft movement
update PROJECT-DESCRIPTION-TEMPLATE.md

perisicMarko committed:
unsuccessful framebuffer implementation

unsuccessful framebuffer implementation

perisicMarko committed:
post review framebuffer imlementation

perisicMarko committed:
implement Framebuffer class
add Framebuffering technique into app(without postprocessing theme)

perisicMarko committed:
make MoonEvent
encapsulate turning moon on/off in MoonEvent class

perisicMarko committed:
Merge branch 'fbuffer' into mi23307

# Conflicts:
#	app/include/MainController.hpp
#	app/src/MainController.cpp
#	engine/include/engine/platform/PlatformController.hpp
#	engine/src/PlatformController.cpp

perisicMarko committed:
add meteor models

perisicMarko committed:
unsuccessful framebuffer postprocessing

perisicMarko committed:
add bind and unbind functions to Framebuffer.hpp class and erase vbo var
add blur rendering to postprocessing shader

perisicMarko committed:
fix draw function

perisicMarko committed:
selective postprocessing setup

perisicMarko committed:
implement bloom

perisicMarko committed:
Encapsulate bloom implementation into Bloom class.
Minor code cleanse.
Update PROJECT-DESCRIPTION-TEMPLATE.md.
Rename bloom/PP shader.

perisicMarko committed:
refactor Bloom and MoonEvent
adjust BrightColor borders

spaske00 commented:
> U template fajlu su bloom i fbuff w pp lekcije iz iste grupe, to sam video tek kada sam zavrsio projekat. Na sajtu stoji da su to lekcije iz razlicite grupe. Sta je predvidjeno za ovu godinu, to iz template-a i ili informacije sa sajta?

Јесу, грешка је темплејту, оно што пише на сајту је добро.
Дакле, блум ефекат је лекција групе Б.

spaske00 commented on PROJECT-DESCRIPTION-TEMPLATE.md:
```diff
@@ -1,30 +1,38 @@
-[ ] Two types of lighting with customizable colors and movement through GUI or ACTIONS
+[x] Model with lighting
+[x] Two types of lighting with customizable colors and movement through GUI or ACTIONS
 [ ] {ACTION_X} --- AFTER_M_SECONDS---Triggers---> {EVENT_A} ---> AFTER_N_SECONDS---Triggers---> {EVENT_B}
```
Ovo je obavezan deo projekta bez kojeg se projekat ne boduje, da li ste i taj deo imlpementiratli? Ako jeste, molim Vas popunite i ovaj deo šablona.

perisicMarko replied:
Event mi je gašenje meseca, odnosno izvora svetlosti. Ne znam da li to ispunjava kriterijum dogadjaja. Obzirom da se cela scena menja, smatram da to jeste dogadjaj.
Hteo sam da unapred da se mesec gasi i pali sa blagim fade-om. Posto ne postoji opcija postavljanja sita za time funkcije da li mislite da tako nesto moze da se odradi elegentno?

spaske00 replied:
> ne postoji opcija postavljanja sita za time funkcije

Nisam siguran šta znači _sit za time funkcije_, ako može pojašnjenje? :) 

> da li mislite da tako nesto moze da se odradi elegentno?

Veoma, FTM (Finite State Machine) će tu biti od pomoći.

spaske00 replied:
> Event mi je gašenje meseca, odnosno izvora svetlosti. Ne znam da li to ispunjava kriterijum dogadjaja. Obzirom da se cela scena menja, smatram da to jeste dogadjaj.  

To bi bio {ACTION_X}, nakon toga na sceni treba da se izdešava niz događaja bez unosa korisnika.

perisicMarko replied:
Mislio sam na "seed" kao seme, malo sam napravio konfuziju. Po secanju na funkcije iz C-a(srand, time) mislio sam da moze da se postigne veca kontrola vrednosti koje dobijam za vreme, pomesao sam.

> > ne postoji opcija postavljanja sita za time funkcije
> 
> Nisam siguran šta znači _sit za time funkcije_, ako može pojašnjenje? :)
> 
> > da li mislite da tako nesto moze da se odradi elegentno?
> 
> Veoma, FTM (Finite State Machine) će tu biti od pomoći.

spaske00 commented on app/config.json:
```diff
@@ -1,11 +1,26 @@
+      },
+      "meteor1": {
+        "path" : "meteor1/meteor.obj"
+      }
```
Koliko mogu videti, ovi modeli fale u repozitorijumu. Ukoliko se preveliki za git, molim Vas postavite ih na gugl drajv na primer i ostavite link u opisu projekta.

perisicMarko replied:
Postavicu ih na drajv.

spaske00 commented on app/include/Bloom.hpp:
```diff
@@ -0,0 +1,19 @@
+#ifndef BLOOM_H
+#define BLOOM_H
+
+#include "Framebuffer.hpp"
```
Ukoliko se u zaglavlju koristi samo pokazivač na neku klasu, onda je dovoljna deklaracija umesto uključivanja celog zaglavlja:
`class Framebuffer`

spaske00 commented on app/include/Bloom.hpp:
```diff
@@ -0,0 +1,19 @@
+
+#include "Framebuffer.hpp"
+
+class Bloom {
```
Bloom ne spada u modul `app`, molim Vas prebacite Bloom i Framebuffer u `engine`.

spaske00 commented on app/include/Bloom.hpp:
```diff
@@ -0,0 +1,19 @@
+#define BLOOM_H
+
+#include "Framebuffer.hpp"
+
```
Fali `namespace`

spaske00 commented on app/include/Bloom.hpp:
```diff
@@ -0,0 +1,19 @@
+public:
+    static bool bloom;
+    static void bloom_color_buffers(Framebuffer* buff);
+    static void bind_bloom_textures(Framebuffer* fbuff);
```
Postoji li neki poseban razlog zbog kojeg su ovo statičke funkcije i promenljiva?
Zar ne bi mogle biti obične funkcije?

spaske00 commented on app/include/Bloom.hpp:
```diff
@@ -0,0 +1,19 @@
+//
+// Created by marko on 2/14/25.
```
Dodati dokumentaciju po uzoru na druge klase i funkcije u modulu `engine`.

spaske00 commented on app/include/Framebuffer.hpp:
```diff
@@ -0,0 +1,30 @@
+//
+// Created by marko on 2/11/25.
```
Dodati dokumentaciju za klase i funkcije po uzoru na modul `engine`.

spaske00 commented on app/include/Framebuffer.hpp:
```diff
@@ -0,0 +1,30 @@
+    void bind();
+    void unbind();
+    void draw_framebuffer_rectangle();
+    void copy_stencil_to_texture();
+    void activate_stencil_texture(int slot);
+    static void enable_stencil_testing();
+    static void disable_stencil_writing();
+    static void stencil_func(std::string func, int ref, unsigned int mask);
+    static void stencil_op(std::string sfail, std::string dfail, std::string dpass);
+    static void stencil_mask(int mask);
```
Da li postoji poseban razlog zbog kojeg su ove funkcije static?
Da li je zaista semantika koja se izražava ovim funkcijama zajednička za sve Framebuffer objekte?

spaske00 replied:
Da li sve ove funkcije zaista pripadaju Framebuffer-u, ili su više vezane za OpenGL? Možda bi imalo smisla razdvojiti ih ako nisu direktno povezane sa samim Framebuffer-om.

spaske00 commented on app/include/MainController.hpp:
```diff
@@ -0,0 +1,58 @@
+};
+
+class MainController : public engine::core::Controller {
+    friend class MoonEvent;
```
Mislim da MoonEvent ne mora pristupati internim delovima `MainControllera`.

spaske00 commented on app/resources/shaders/bloom_shader.glsl:
```diff
@@ -0,0 +1,66 @@
+    vec3 result = screenColor + bloomSwitch * col;
+    //const float gamma = 2.2;
+    //result = result - exp(-screenColor * exposure);
+    //result = pow(result, vec3(1.0 / gamma));
```
Nije dobro ostavljati zakomentarisan kod. Ako radi, otkomentarisati, ako ne radi obrisati.

Deluje mi da će biti potreban zbog podešavanja `exposure` parametra.

spaske00 commented on app/resources/shaders/meteor.glsl:
```diff
@@ -0,0 +1,121 @@
+in vec3 FragPos;
+in vec3 Normal;
+
+struct Light{
```
Možda će doprineti jasnoći koda ako se struktura preimenuje u PointLight.

spaske00 replied:
Takođe, fale parametri za računanje slabljenja svetlosti.

spaske00 commented on app/resources/shaders/meteor.glsl:
```diff
@@ -0,0 +1,121 @@
+    vec3 intensity;
+};
+
+struct SpotLight{
```
Fale parametri za računanje slabljenja svetlosti.

spaske00 commented on app/resources/shaders/meteor.glsl:
```diff
@@ -0,0 +1,121 @@
+    diffuse += diffuse_spotLight;
+    specular += specular_spotLight;
+    vec3 result = ambient + diffuse + specular;
+
```
Malo je teže pratiti kod kada je računanje obe vrste osvetljenja u istoj funkciji. Mislim bi bilo preglednije ako bi se računanje različitih tipova osvetljenja izmestilo u zasebne funkcije.

spaske00 commented on app/resources/shaders/moon.glsl:
```diff
@@ -0,0 +1,40 @@
+void main(){
+    vec3 result = light_intensity * texture(texture_diffuse1, TexCoords).rgb;
+    float brightness = dot(result, vec3(0.2126, 0.7152, 0.0722));
+        if(brightness > 0.9)
```
Poravnati if-ove.

spaske00 commented on app/resources/shaders/skybox.glsl:
```diff
@@ -0,0 +1,34 @@
+void main(){
+    vec3 result = texture(skybox, TexCoords).rgb;
+    float brightness = dot(result, vec3(0.2126, 0.7152, 0.0722));
+            if(brightness > 1.0)
```
Poravnati if-ove

spaske00 commented on app/resources/shaders/space_craft.glsl:
Isti komentari kao za `meteor.glsl`

spaske00 commented on app/resources/shaders/space_station.glsl:
Isti komentari kao za `meteor.glsl`

spaske00 commented on app/src/Bloom.cpp:
```diff
@@ -0,0 +1,23 @@
+// Created by marko on 2/14/25.
+//
+
+#include "../include/Bloom.hpp"
+#include <../../engine/include/engine/graphics/OpenGL.hpp>
+#include <../../engine/include/engine/platform/PlatformController.hpp>
+#include "../include/Framebuffer.hpp"
```
Nisu potrebne relativne putanje od trenutnog fajla: `../`. Dovoljno je 
`engine/graphics/OpenGL.hpp`
`Framebuffer.hpp`

spaske00 commented on app/src/Framebuffer.cpp:
```diff
@@ -0,0 +1,66 @@
+
+
+Framebuffer::Framebuffer() {
+    auto platform = engine::core::Controller::get<engine::platform::PlatformController>();
+
+    int width = platform->window()->width();
+    int height = platform->window()->height();
```
Isti komentar kao: [https://github.com/marko11120/matf-rg-project-2024/pull/3#discussion_r1950362758](https://github.com/marko11120/matf-rg-project-2024/pull/3#discussion_r1950362758)

spaske00 commented on engine/src/OpenGL.cpp:
```diff
@@ -203,4 +216,186 @@ namespace engine::graphics {
+    void OpenGL::disable_stencil_writing() {
+        CHECKED_GL_CALL(glStencilMask, 0x00);
+    }
+    void OpenGL::stencil_func(std::string func, int ref, unsigned int mask) {
```
Mislim da bi umesto stringa bilo zgodnije napraviti enum sa mogućim vrednostima.

Ipak stencil funkcija ne može biti bilo koji string.

spaske00 commented on engine/src/OpenGL.cpp:
```diff
@@ -203,4 +216,186 @@ namespace engine::graphics {
+
+        CHECKED_GL_CALL(glStencilFunc, func_, ref, mask);
+    }
+    void OpenGL::stencil_op(std::string sfail, std::string dfail, std::string dpass) {
```
Isti komentar kao iznad.

perisicMarko replied:
Razresio sam komentar iznad, ako je bilo za to da se ovo poziva vise puta to nije slucaj. Funkcija se poziva u konstruktoru Framebuffer-a koji se poziva prilikom funkcije initialize u MainController-u.

spaske00 commented on engine/src/OpenGL.cpp:
```diff
@@ -203,4 +216,186 @@ namespace engine::graphics {
+    }
+
+    unsigned int OpenGL::copy_stencil_to_texture() {
+        auto platform = engine::core::Controller::get<engine::platform::PlatformController>();
+        int width     = platform->window()->width();
+        int height    = platform->window()->height();
```
Isti komentar kao: [https://github.com/marko11120/matf-rg-project-2024/pull/3#discussion_r1950362758](https://github.com/marko11120/matf-rg-project-2024/pull/3#discussion_r1950362758)

spaske00 commented on engine/src/OpenGL.cpp:
```diff
@@ -203,4 +216,186 @@ namespace engine::graphics {
+        auto platform = engine::core::Controller::get<engine::platform::PlatformController>();
+        int width     = platform->window()->width();
+        int height    = platform->window()->height();
+        unsigned int stencil_texture;
+        CHECKED_GL_CALL(glGenTextures, 1, &stencil_texture);
+        CHECKED_GL_CALL(glBindTexture, GL_TEXTURE_2D, stencil_texture);
```
Čini mi se da ovaj pristup možda nije dovoljno fleksibilan ni efikasan, jer će u svakom frejmu biti kreiran novi OpenGL objekat.

Možda bi bilo bolje omogućiti prosleđivanje src i dst kopiranja kao argumente ove funkcije, slično kao što to radi memcpy. To bi moglo poboljšati efikasnost i omogućiti veću kontrolu nad procesom kopiranja

spaske00 commented on engine/src/OpenGL.cpp:
```diff
@@ -203,4 +216,186 @@ namespace engine::graphics {
+        CHECKED_GL_CALL(glTexImage2D, GL_TEXTURE_2D, 0, GL_R8, width, height, 0, GL_RED, GL_UNSIGNED_BYTE, nullptr);
+
+        // vec for storing buff data
+        std::vector<GLubyte> stencilData(width * height);
```
Za kopiranje vrednosti između OpenGL objekata nije potrebna virtuelna memorija procesa. Potražite odgovarajuću `glCopy` funkcija iz dokumentacije [https://docs.gl/](https://docs.gl/)

spaske00 commented on engine/src/OpenGL.cpp:
```diff
@@ -203,4 +216,186 @@ namespace engine::graphics {
+        CHECKED_GL_CALL(glBindTexture, GL_TEXTURE_2D, texture);
+    }
+
+    void OpenGL::bloom_color_buffers(unsigned int buffers[], int SCR_WIDTH, int SCR_HEIGHT) {
```
U engine modulu, nazivi parametara se obično pišu malim slovima, pa bi bilo dobro uskladiti se s tom konvencijom.

Takođe, naziv funkcije i njena svrha deluju malo neusklađeno – funkcija zapravo kreira bafer, a obično funkcije koje nešto stvaraju počinju sa create, new ili make, što bi moglo učiniti namenu jasnijom.

Još jedno pitanje – da li ovo mora biti ovako usko vezano za bloom? Izgleda da se ovde kreiraju dva ista framebafera, pa možda ima smisla napraviti generičku funkciju koja kreira bafer sa color attachment-om i pozvati je dva puta umesto dupliranja koda?

spaske00 commented on engine/src/PlatformController.cpp:
```diff
@@ -281,4 +280,20 @@ namespace engine::platform {
     }
 
+    void PlatformController::set_cursor_visible(bool flag) {
+        auto platform = engine::core::Controller::get<platform::PlatformController>();
```
Pošto se ova funkcija već nalazi unutar instance PlatformController, čini mi se da bi bilo dovoljno koristiti window() umesto platform->window(). To bi učinilo kod čitljivijim i malo pojednostavilo pristup.

spaske00 commented on app/include/Framebuffer.hpp:
```diff
@@ -0,0 +1,30 @@
+#define FRAMEBUFFER_H
+#include <string>
+
+class Framebuffer {
```
Čini mi se da bi Framebuffer trebalo da bude deo engine modula, kako bi ostao opšte upotrebljiv i odvojen od specifičnih implementacija.

spaske00 commented on app/include/Framebuffer.hpp:
```diff
@@ -0,0 +1,30 @@
+#include <string>
+
+class Framebuffer {
+    friend class Bloom;
```
Izgleda da Bloom trenutno ima pristup internim delovima klase Framebuffer. Možda bi bilo bolje razmotriti drugačiju organizaciju kako bi se poboljšala enkapsulacija i smanjila zavisnost između modula?

spaske00 commented on app/include/MainController.hpp:
```diff
@@ -0,0 +1,58 @@
+    friend class MoonEvent;
+    friend class MainPlatformEventObserver;
+public:
+    MoonEvent* moon_event_handler;
+    float exposure = 1;
+    glm::vec3 spacecraft_pos = glm::vec3(0.0f, -2.0f, -30.0f);
+    float spacecraft_rotation = glm::radians(0.f);
```
Da li se ovim poljima zaista pristupa iz neke druge klase spolja? Ako je to slučaj, možda vredi razmotriti da li je taj pristup zaista neophodan. Ako nije, možda bi bilo bolje postaviti ih kao private radi bolje enkapsulacije.

spaske00 commented on app/include/MoonEvent.hpp:
```diff
@@ -0,0 +1,21 @@
+    MoonEvent();
+    glm::vec3 turn_on();
+    glm::vec3 turn_off();
+    int moon_event;
```
Gledajući ovu klasu, nije potpuno jasno zašto turn_on i turn_off vraćaju vec3 i šta tačno taj vec3 predstavlja.

Obično, metode turn_on i turn_off sugerišu da se neki proces pokreće ili zaustavlja i da zatim radi u pozadini, bez potrebe da pozivalac dalje upravlja njime. Možda bi bilo korisno razmotriti da li je ovakvo ponašanje optimalno ili bi se moglo jasnije modelovati?

spaske00 commented on app/src/Bloom.cpp:
```diff
@@ -0,0 +1,23 @@
+
+void Bloom::bloom_color_buffers(Framebuffer* fbuff) {
+    auto platform = engine::core::Controller::get<engine::platform::PlatformController>();
+    int width = platform->window()->width();
```
Isti komentar kao za [https://github.com/marko11120/matf-rg-project-2024/pull/3#discussion_r1950362758]( https://github.com/marko11120/matf-rg-project-2024/pull/3#discussion_r1950362758)

spaske00 commented on app/src/Framebuffer.cpp:
```diff
@@ -0,0 +1,66 @@
+    engine::graphics::OpenGL::draw_arrays(6);
+}
+
+void Framebuffer::enable_stencil_testing() {
+    engine::graphics::OpenGL::enable_stencil_testing();
+}
+void Framebuffer::disable_stencil_writing() {
+    engine::graphics::OpenGL::disable_stencil_writing();
+}
+
+void Framebuffer::stencil_func(std::string func, int ref, unsigned int mask) {
+    engine::graphics::OpenGL::stencil_func(func, ref, mask);
+}
+void Framebuffer::stencil_op(std::string sfail, std::string dfail, std::string dpass) {
+    engine::graphics::OpenGL::stencil_op(sfail, dfail, dpass);
+}
+void Framebuffer::stencil_mask(int mask) {
+    engine::graphics::OpenGL::stencil_mask(mask);
+}
```
Čini mi se da bi ove funkcije mogle biti suvišne, jer samo prosleđuju sve parametre direktno engine::graphics::OpenGL funkcijama, bez dodatne logike.

Možda bi bilo jednostavnije i preglednije pozivati engine::graphics::OpenGL direktno, osim ako postoji neki razlog za ovakav posrednički sloj?

spaske00 commented on app/src/MyApp.cpp:
```diff
@@ -0,0 +1,12 @@
+#include <engine/platform/PlatformController.hpp>
+
+#include "../include/MyApp.hpp"
```
`#include <MyApp.hpp`

spaske00 commented on app/src/MyApp.cpp:
```diff
@@ -0,0 +1,12 @@
+
+#include "../include/MyApp.hpp"
+
+#include "../../engine/libs/spdlog/include/spdlog/spdlog.h"
```
`#include <spdlog/spdlog.h>`

spaske00 commented on engine/include/engine/graphics/OpenGL.hpp:
```diff
@@ -132,6 +133,27 @@ namespace engine::graphics {
+
+
+        static unsigned int create_framebuffer();
+        static unsigned int create_color_attachment(int SCR_WIDTH, int SCR_HEIGHT);
+        static unsigned int create_render_buffer(int SCR_WIDTH, int SCR_HEIGHT);
```
U skladu s konvencijama koda, imena parametara bi trebalo da budu pisana malim slovima radi doslednosti.

spaske00 commented on engine/include/engine/graphics/OpenGL.hpp:
```diff
@@ -132,6 +133,27 @@ namespace engine::graphics {
+        static void blend();
+        static void clear_color(float a, float b, float c, float d);
+        static void bind_buffer(unsigned int bufferId);
+        static void draw_arrays(int x);
```
Nije potpuno jasno šta x predstavlja. Možda bi deskriptivniji naziv pomogao boljoj čitljivosti koda?

spaske00 commented on engine/include/engine/graphics/OpenGL.hpp:
```diff
@@ -132,6 +133,27 @@ namespace engine::graphics {
+        static unsigned int create_render_buffer(int SCR_WIDTH, int SCR_HEIGHT);
+        static unsigned int configure_framebuffer_rectangle();
+        static void bind_framebuffer(unsigned int framebuffer);
+        static void bind_colorbuffer(unsigned int colorbuffer);
```
Da li bind_colorbuffer praktično radi isto što i aktiviranje teksture? Ako da, možda bi bilo korisno razmotriti da li je potrebno zadržati oba naziva ili ih ujediniti radi jasnije organizacije koda.

spaske00 commented on engine/include/engine/graphics/OpenGL.hpp:
```diff
@@ -132,6 +133,27 @@ namespace engine::graphics {
+        static void bind_colorbuffer(unsigned int colorbuffer);
+        static void blend();
+        static void clear_color(float a, float b, float c, float d);
+        static void bind_buffer(unsigned int bufferId);
```
`buffer_id`

spaske00 commented on engine/include/engine/graphics/OpenGL.hpp:
```diff
@@ -132,6 +133,27 @@ namespace engine::graphics {
+        static void bind_framebuffer(unsigned int framebuffer);
+        static void bind_colorbuffer(unsigned int colorbuffer);
+        static void blend();
+        static void clear_color(float a, float b, float c, float d);
```
Možda bi imena parametara mogla biti r, g, b, a, s obzirom na to da predstavljaju komponente boje? To bi moglo poboljšati čitljivost i jasnoću koda.

spaske00 commented on engine/include/engine/graphics/OpenGL.hpp:
Bilo bi korisno detaljno dokumentovati dodate funkcije po uzoru na već postojeći kod, kako bi konzistentnost i čitljivost bile očuvane. Na taj način će buduće izmene i održavanje biti lakši.

spaske00 commented on engine/src/Camera.cpp:
```diff
@@ -1,5 +1,7 @@
 #include <engine/graphics/Camera.hpp>
+#include <engine/platform/PlatformController.hpp>
+
+#include "GLFW/glfw3.h"
```
#include <GLFW/glfw3.h>

spaske00 commented on engine/src/Camera.cpp:
```diff
@@ -1,5 +1,7 @@
-
 #include <engine/graphics/Camera.hpp>
+#include <engine/platform/PlatformController.hpp>
```
"Da li je ovo zaglavlje zaista potrebno? Ako se ne koristi, možda bi ga bilo bolje ukloniti radi čistijeg i urednijeg koda.

spaske00 commented on engine/src/GraphicsController.cpp:
Prazne ili obrisane linije mogu dodati nepotreban šum u PR-u, jer uključuju fajlove u kojima suštinski nije bilo promena. Možda bi bilo korisno obratiti pažnju na to kako bi PR ostao čist i fokusiran na relevantne izmene.

spaske00 commented on engine/src/OpenGL.cpp:
```diff
@@ -1,13 +1,18 @@
 #include <engine/resources/Skybox.hpp>
 #include <engine/util/Errors.hpp>
 #include <engine/util/Utils.hpp>
+#include <GL/gl.h>
```
Nepotrebno zaglavlje.

perisicMarko replied:
To zaglavlje je ostalo od pokusaja implementacije renderovanja u stencil buffer, obrisacu.

spaske00 commented on engine/src/OpenGL.cpp:
```diff
@@ -1,13 +1,18 @@
 #include <engine/util/Utils.hpp>
+#include <GL/gl.h>
+#include <array>
+#include <filesystem>
```
Nepotrebno zaglavlje.

spaske00 commented on engine/src/ResourcesController.cpp:
Prazne ili obrisane linije mogu dodati nepotreban šum u PR-u, jer uključuju fajlove u kojima suštinski nije bilo promena. Možda bi bilo korisno obratiti pažnju na to kako bi PR ostao čist i fokusiran na relevantne izmene.

spaske00 commented on engine/test/app/src/MainController.cpp:
Prazne ili obrisane linije mogu dodati nepotreban šum u PR-u, jer uključuju fajlove u kojima suštinski nije bilo promena. Možda bi bilo korisno obratiti pažnju na to kako bi PR ostao čist i fokusiran na relevantne izmene.

spaske00 reviewed (changes requested):
Ostavio sam komentare samo za kod.
Primetio sam da su meteor.glsl, space_craft.glsl, i space_station.glsl praktično identični. 
Da li propuštam neke sitne varijacije ili bi sva tri mogli biti zamenjeni jednim šejderom?



U nastavku greška koja se dobije prilikom pokretanja programa:
```
/home/ms/matf/rg/projekti/mi23307_marko11120/app/APP
[2025-02-15 12:56:25.821] [info] ArgParser initialized.
[2025-02-15 12:56:25.821] [info] Configuration initialized.
[2025-02-15 12:56:25.821] [info] Hello, setup!
[2025-02-15 12:56:25.821] [info] EngineControllersBegin::initialize
[2025-02-15 12:56:25.821] [info] PlatformGLFW3Controller::initialize
[2025-02-15 12:56:26.100] [info] Platform[GLFW 3.5.0]
[2025-02-15 12:56:26.100] [info] GraphicsController::initialize
[2025-02-15 12:56:26.101] [info] ResourcesController::initialize
[2025-02-15 12:56:26.101] [info] load_shader(path=resources/shaders/skybox.glsl)
[2025-02-15 12:56:26.145] [info] load_shader(path=resources/shaders/moon.glsl)
[2025-02-15 12:56:26.157] [info] load_shader(path=resources/shaders/space_station.glsl)
[2025-02-15 12:56:26.182] [info] load_shader(path=resources/shaders/space_craft.glsl)
[2025-02-15 12:56:26.205] [info] load_shader(path=resources/shaders/meteor.glsl)
[2025-02-15 12:56:26.228] [info] load_shader(path=resources/shaders/bloom_shader.glsl)
[2025-02-15 12:56:26.234] [info] load_model(name=Moon, path=resources/models/Moon/Moon.obj)
[2025-02-15 12:56:26.844] [info] load_texture(path=resources/models/Moon/Moon.jpg)
[2025-02-15 12:56:28.118] [info] load_model(name=meteor1, path=resources/models/meteor1/meteor.obj)
[2025-02-15 12:56:28.119] [error] AssetLoadingError: in /home/ms/matf/rg/projekti/mi23307_marko11120/engine/src/ResourcesController.cpp:136.
 Message: Assimp error while reading model: resources/models/meteor1/meteor.obj from path meteor1.
[2025-02-15 12:56:28.119] [info] 14MainController::terminate
[2025-02-15 12:56:28.119] [info] EngineControllersEnd::terminate
[2025-02-15 12:56:28.119] [info] ResourcesController::terminate
[2025-02-15 12:56:28.119] [info] GraphicsController::terminate
[2025-02-15 12:56:28.167] [info] PlatformGLFW3Controller::terminate
[2025-02-15 12:56:28.167] [info] EngineControllersBegin::terminate

Process finished with exit code 0

```

perisicMarko commented:
Razlika u sejderima su granicne vrednosti za to da li fragment ide u bright color buffer, ali to mogu resiti uniformnom promenljivom. Izmenicu da bude jedan sejder za stanicu, s brod i meteore. Sejder meseca ne implementira svetlo tako da njega ne mogu da izostavim.

Nije mi bas jasno zašto imate ovu grešku, trebalo bi da je metero1 folder push-an na git. Provericu šta se desilo kada budem pri laptopu, tada cu napraviti i izmene koje prate komentare.

spaske00 commented:
> Razlika u sejderima su granicne vrednosti za to da li fragment ide u bright color buffer, ali to mogu resiti uniformnom promenljivom. Izmenicu da bude jedan sejder za stanicu, s brod i meteore. Sejder meseca ne implementira svetlo tako da njega ne mogu da izostavim.

Mislim da će tako biti najbolje. Ako nema mnogo razlika uvek ih možete tako objediniti.

perisicMarko committed:
refactor shaders

perisicMarko committed:
refactor Bloom, Framebuffer, OpenGL

perisicMarko commented:
Pushao sam promene, nadam se da nisam nesto propustio. 
Ostaje dogadjaj, fade meseca i da sredim linkove.

perisicMarko committed:
add Bloom.cpp, Framebuffer.hpp
implement FSM functionality for moon
write brief for added functions in OpenGL.hpp

perisicMarko committed:
implement actions
update PROJECT-DESCRIPTION-TEMPLATE.md

spaske00 commented:
Poštovani kolega,
Molim Vas, nemojte zatvarati [Review] PR. Sve komitove ispravki postavite na ovoj grani. Kada ispravite sve komentare, javite mi da pogledam ponovo.
Molim Vas, ponovo otvorite ovaj PR i obrišite novi PR https://github.com/marko11120/matf-rg-project-2024/pull/6 i ako imate promena na toj grani, postavite ih ovde.

perisicMarko commented:
Propustio sam ovu poruku, unsubscribe-ovao sam se kada sam zatvorio PR.
Sa poslednja dva commit-a na ovoj grani sam zavrsio sa ispravkama, dopunio implementaciju dogadjaja i fade meseca, tako da se sve nalazi na na ovoj grani. Takodje sam dodao kratke deskripcije funkcija koje sam dodao u OpenGL klasu.

perisicMarko committed:
correct event night light

spaske00 reviewed (changes requested):
Prelistao sam promene i primetio da se neke ranije greške ponavljaju, poput onih u engine/src/Bloom.cpp.Takođe na mnogim novododanim funkcijama i klasama i dalje nedostaje dokumentacija.

Molim Vas da još jednom prođete kroz ceo PR, pregledate sve prethodne komentare koje smo ostavili i proverite kako se oni odnose na najnovije commit-ove. Bilo bi dobro da kod uskladite sa tim sugestijama.

Javite mi kada završite, da mogu da pregledam.

perisicMarko committed:
post review corrections

perisicMarko committed:
fix name of Bloom function called in MainController

perisicMarko commented:
Ispravke:
-Iskomentarisao sam sve sto nije override-ovano iz naslednje funkcije.
-Obrisao sam nepotrebne prefix-e u Bloom i Framebuffer klasama.
-Uskladio imena funkcija koje su mi promakle iz prethodnih komentara.
-Obrisao nepotreban kod.

perisicMarko committed:
complete functions documentation
refactor Bloom, Framebuffer functions
fix flip
fix include paths and erase unnecessary includes

spaske00 commented on PROJECT-DESCRIPTION-TEMPLATE.md:
```diff
@@ -1,45 +1,51 @@
-# {NAME-OF-THE-PROJECT}
+# {Space station}
```
Ukloniti {} svuda gde su podaci popunjeni, zagrade su deo šablona.

spaske00 commented on PROJECT-DESCRIPTION-TEMPLATE.md:
```diff
@@ -1,45 +1,51 @@
-[ ] Model with lighting
-[ ] Two types of lighting with customizable colors and movement through GUI or ACTIONS
-[ ] {ACTION_X} --- AFTER_M_SECONDS---Triggers---> {EVENT_A} ---> AFTER_N_SECONDS---Triggers---> {EVENT_B}
+[x] Model with lighting
+[x] Two types of lighting with customizable colors and movement through GUI or ACTIONS
+[x] {Moon on/off} --- AFTER_M_SECONDS---Triggers---> {Moon position change} ---> AFTER_N_SECONDS---Triggers---> {space station night light on}
```
Zameniti i ostale stvake konkretnim opisom onoga što se nalazi u projektu.
Zameniti AFTER_M/N_SECONDS takođe

spaske00 commented on PROJECT-DESCRIPTION-TEMPLATE.md:
```diff
@@ -1,45 +1,51 @@
-[link2]
+[https://drive.google.com/drive/folders/1g7tulu0S9fIwor31Qgwyl0m8sWO1cCXX?usp=sharing]
 
 ## Textures
```
Srediti i ostatak opisa, ako nema linkova do resursa obrisati šta je višak od naslova.

spaske00 commented on app/resources/shaders/space_craft.glsl:
```diff
@@ -0,0 +1,119 @@
+in vec2 TexCoords;
+in vec3 FragPos;
+
+struct Light{
```
Fale parametri za slabljenje svetlonsti

perisicMarko replied:
Taj shader više nije deo projekta, integrisao sam ga u space_object.glsl. Treba da ga sklonim i sa githuba.

spaske00 commented on app/resources/shaders/space_craft.glsl:
```diff
@@ -0,0 +1,119 @@
+};
+
+
+struct SpotLight{
```
Fale parametri za slabljenje svetlosti

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,389 @@
+    auto platform = engine::core::Controller::get<engine::platform::PlatformController>();
+    platform->set_cursor_visible(false);
+
+    m_framebuffer = new engine::graphics::Framebuffer();
+    m_moon_event_handler           = new MoonEvent();
```
Ako se ovi objekti ne obrišu ekslicitno onda je ovo curenje memorije. 
Molim Vas, koristite std::unique_ptr u ovom slučaju.

spaske00 reviewed (commented):
Scena izgleda odlično, a događaj funkcioniše kako treba!

Imam nekoliko manjih zapažanja:

Bloom efekat bi mogao biti izraženiji na svemirskoj stanici.
Lampa trenutno ne obasjava mesec.
Ovo su sitnije izmene, pa ih možete doraditi kada budete imali vremena.

Što se tiče lekcije Framebuffer, primetio sam da je štiklirana, ali nisam primetio dodatno postprocesiranje scene nakon Bloom efekta.

Bloom efekat i postprocesiranje su dve odvojene lekcije. Blur efekat u sklopu Bloom-a jeste sastavni deo Bloom lekcije, ali se on primenjuje samo na svetle fragmente, a ne na celu scenu.

Da bi lekcija iz grupe A Framebuffer i postprocessing bila kompletna, potrebno je dodati još jedan sloj postprocesiranja scene. Mislim da bi, na primer, primena grey-scaling efekta kada se aktivira MoonEvent mogla lepo da se uklopi i obogati scenu.

Ako vam treba dodatna pomoć oko postprocesiranja ili imate neka pitanja, slobodno se javite! Sve u svemu, projekat lepo napreduje.

perisicMarko committed:
add greyscale PP effect
implement moon spotlight
override terminate function(fix memory leak)
update PROJECT-DESCRIPTION-TEMPLATE.md

perisicMarko commented:
Doterao sam projekat po sugestijama koje ste naveli. 
Imao sam problem sa unique ptr jer sam koristio forward deklaraciju u MainController klasi za pokazivace, a on mi je valjda trazio celu definiciju klase te sam predefinisao terminate i pozvao default destruktore za dve instance koje sam cuvao u MC-u.
Uklopio sam greyscale u scenu sa formulom, hteo sam da dodam neko PP ali nisam znao sta. A obzirom da je isti princip kao za blur mislio sam da je ok, pa sam zato taj deo izostavio.
Zaboravio sam da dodam spotlight na mesec nakon implementacije dogadjaja, pre toga mi nije bilo smisleno.
Takodje sam se poigrao malo sa granicama za bloom. Bloom je bio dobar dok su strukture bile naopacke :). 

Ako je sada sve u redu sa projektom, hteo bih da Vas zamolim da me uputite na neku literaturu za pisanje cistog koda. Vratili ste me dosta puta zbog nepreglednog i dvosmislenog koda te vidim da Vam je stalo da se studenti kod Vas na kursu unaprede i na tom polju. Da li bi klasik kao "Clean Code" bio dobar pocetak ili postoji savremenija alternativa.

perisicMarko commented:
P.S. Space craft shader mi stoji i dalje u projektu, nisam siguran kako da ga obrisem. Lokalno sam ga obrisao pa mi nije jasno zasto nije star sadrzaj direktorijuma shaders pregazen kada sam pushao trenutno stanje.

spaske00 commented:
> Imao sam problem sa unique ptr jer sam koristio forward deklaraciju u MainController klasi za pokazivace, a on mi je valjda trazio celu definiciju klase te sam predefinisao terminate i pozvao default destruktore za dve instance koje sam cuvao u MC-u.

Da, klasi `std::uniqe_ptr<T>` je potrebna cela definicija klase jer upravlja životnim vekom objekta tipa `T`. U tom slučaju, slobodno uključite celo zaglavlje `Framebuffer.hpp`. 

> Takodje sam se poigrao malo sa granicama za bloom. Bloom je bio dobar dok su strukture bile naopacke :).  

Sada nekako baš blješte, pogotovu kada se obasjaju sa lampom :/

> Ako je sada sve u redu sa projektom

Na obrise mi deluje da ima u kodu još toga što bi moglo da se utegne, ali projekat ste sigurno položili sa dobrim bodovima.  
Preporučio bih Vam da napravite pauzu od par dana od projekta pa da se vratite kada se sve slegne i još jednom probate sami da prođete kroz sve naše prethodne komentare, pa da mi javite onda da pogledam još jednom.

> ... hteo bih da Vas zamolim da me uputite na neku literaturu za pisanje cistog koda. Vratili ste me dosta puta zbog nepreglednog i dvosmislenog koda te vidim da Vam je stalo da se studenti kod Vas na kursu unaprede i na tom polju. Da li bi klasik kao "Clean Code" bio dobar pocetak ili postoji savremenija alternativa. 

Meni se "A Philosophy of Software Design - John Ousterhout" više dopala. Svaki jezik ima neke svoje specifičnosti i način pisanja "idiomatskog" koda, C++ je tu možda najzahtevniji. Imajte u vidu da se i veći deo veštine pisanja koda stekne iskustvom, u knjigama je zabeležen samo jedan mali deo nečijeg iskustva.

> P.S. Space craft shader mi stoji i dalje u projektu, nisam siguran kako da ga obrisem. Lokalno sam ga obrisao pa mi nije jasno zasto nije star sadrzaj direktorijuma shaders pregazen kada sam pushao trenutno stanje.

Za obrisane fajlove treba uraditi `git rm` umesto `git add` kako bih sistem obirsao iz evidencije.

perisicMarko committed:
Delete app/resources/shaders/space_craft.glsl

shader integrated into space_objects.glsl

perisicMarko committed:
cleanse code

perisicMarko commented:
Napravio sam neke sitnije izmene u projektu, ne znam da li sam promasio temu.  Da li bi ovako uradjen projekat osvojio preko 30 poena? 
Hvala za preporuku knjige :)

perisicMarko committed:
fix var names

perisicMarko committed:
make spotlight available when moon off
make switch for spotlight
update PROJECT-DESCRIPTION-TEMPLATE.md

perisicMarko commented:
Sa ovim dodatnim izmenama mislim da sam zavrsio. Izmenio sam postojanje spotlight-a na sceni jer mi onako inicijalno postavljen nema smisla, pogotovo jer sam dodao bloom.

spaske00 commented on engine/src/PlatformController.cpp:
```diff
@@ -281,14 +281,13 @@ namespace engine::platform {
     }
 
     void PlatformController::set_cursor_visible(bool flag) {
```
[REFACTOR] Ovo je funkcija `PlatformController::set_enable_cursor`. Mislim da bi Vam bilo lakše da ste samo nju proširili sa čuvanjem stanja `m_cursor_visible`.

perisicMarko replied:
Ovo sam refaktorisao, samo sam zaboravio da razresim komentar nakon sto sam pushao kod koji implementira promene.

spaske00 commented on engine/src/OpenGL.cpp:
```diff
@@ -203,4 +206,193 @@ namespace engine::graphics {
+    void OpenGL::copy_stencil_to_texture(int width, int height, unsigned int stencil_texture) {
+
+        // copy to texture from stencil buff
+        auto stencil_data = new GLubyte[width * height];
```
[CRITICAL] Ova linija će prouzrokovati curenje memorije u svakom frejmu. Ključna reč auto u ovom kontekstu znači da će kompajler dedukovati tip promenljive stencil_data na osnovu izraza sa desne strane =, ali to ne znači da će memorija biti automatski oslobođena.

Baš za ovu funkciju sam preporučio korišćenje neke od glCopy funkcija kako bi se izbegla dodatna alokacija memorije (videti komentar: [link](https://github.com/marko11120/matf-rg-project-2024/pull/5/files/50c87079f34e59ac66dda5f0675909a9c768ba61#r1957116937)).

Vidim da je komentar označen kao rešen, ali nije ostavljen odgovor niti je predložena promena implementirana. Da bismo bili efikasniji u pregledanju i davanju povratnih informacija, molim Vas da:
- Implementirate predlog ako je primenljiv,
- Ako niste sigurni, slobodno postavite pitanje,
- Ako predlog nije implementiran, nemojte ga označavati kao rešen dok se ne razjasni.

Ovo smo već napomenuli u [ovom komentaru](https://github.com/marko11120/matf-rg-project-2024/pull/3#issuecomment-2653744294), pa bih voleo da se ubuduće držimo ove prakse kako bismo olakšali dalji rad.

perisicMarko replied:
Svakako funkcija nije ni radila ono sto treba, kopirao sam prazan niz u pomocnu texturu. Gledam opet dokumentaciju i ja ne vidim ni jednu funkciju koja radi ono sto meni treba, da kopira iz stencil buffer-a u teksturu, samo iz color i depth buffera. Jedino glCopyPixels pominje stencil buffer ali nisam siguran da sam skapirao sta ona tacno radie. Ovaj deo oko stencil buffer-a mi i nije bitan za projekat, odustao sam na kraju od selektivnog PP jer  sam  presao na implementaciju bloom-a i ocekivao da cu tu da dodam neki efekat celoj sceni.(Momenat gde sam mislio da se blur svetlih fragmenata racuna  kao PP). Tako da nigde ne pozivam funkcije koje rade sa stencil buffer-om. Mozda bih mogao i da ih obrisem jer nemam potrebu za njima?

spaske00 replied:
Ako se nigde ne koristi, slobodno obrišite. 

Može i preko niza kao pomoćnog bafera, samo je bitno da memorija bude oslobođena.

spaske00 reviewed (changes requested):
Ostao je još jedan [CRITICAL] komentar koji je potrebno razrešiti.

Trenutni skor projekta je 25/35 bodova. Bodovi su podeljeni tako da polovina dolazi iz implementacije, a polovina iz vizuelnog izgleda scene.

Što se tiče implementacije, bilo je dosta pomoći s moje strane, pa je zbog toga ukupan broj bodova nešto niži. Međutim, sama scena izgleda odlično!

Nadam se da vam je ovaj projekat doneo korisna znanja i iskustva koja će vam koristiti u budućem radu. Ako imate bilo kakva pitanja ili vam je potrebna dodatna pomoć, slobodno se obratite!

perisicMarko committed:
fix copy function(memory leak, emtpy src copy)

perisicMarko commented:
Projekat mi je svakako zanimljiviji od pismenog ispita u bilo kakvoj formi, takodje kompletan proces mi je doneo znanja sto se tice rada na projektu. Poceo sam da razvijam web aplikaciju tako da eto i na vise nivoa je bilo produktivno iskustvo, pogotovo mislim na Vas odziv sto se tice stvari koje nisu vezanu za grafiku. 

Sto se tice bodova, racunao sam  da sam dobio 15/15 bodova za osnovni nivo, dali ste mi hint za FSM ali sam sam istrazio i uklopio pattern u event. Ostalih 10 bi onda doslo iz mogucih 20 na 2 napredne lekcije. Posto je  je podela bodova pola implementacija pola izgled scene, da li to onda znaci da maltene nisam dobio nista na implementaciju ako kazete da scena izgleda odlicno pa verujem da odatle ide 5 poena na obe lekcije. 

P.S. 
Slazem sa da je asistencija sa Vase strane bila uveliko prisutna. 
Samo me zanima jer sam ocekivao vise, nije mi bitna ocena zbog proseka.

spaske00 commented:
Naravno, bodovanje nije nikakva tajna! 🙂

U svakom delu projekta (osnovni deo – 15 bodova, lekcije A i B – po 10 bodova) nedostaje po malo bodova, delom zbog pomoći sa moje strane tokom implementacije, a delom zbog segmenata koje je još potrebno doraditi.

Najveći uticaj na bodove imala je činjenica da se isti tip greške ponavljao između revizija. Kada biste je ispravili na jednom mestu, ponovo bi se pojavila negde drugde ili ne bi bila ispravljena između revizija. Zato bih preporučio da u sledećim iteracijama ili projektu na nekom drugom predmetu/poslu posvetite dodatnu pažnju konzistentnosti ispravki kroz ceo projekat.

Ukoliko želite da doradite projekat i osvojite više bodova, uvek možete implementirati još jednu lekciju iz grupe B i integrisati je u trenutni projekat.

perisicMarko commented:
Razmotricu opciju da unapredim projekat nekom lekcijom iz grupe B, svakako ima vremena i samouveren sam da cu sa stecenim znanjem uklopiti to kako treba ako se odlucim za isto. Takodje cu razmisliti koji su to segmenti koje je potrebno dodatno doraditi.

P.S.Znam da su se greske ponavljale, hvala na strpljenju, malo sam bio preplavljen novim konceptima.

perisicMarko committed:
implement deferred shading technique
refactor MainController.cpp accordingly

perisicMarko commented:
Implementirao sam jos jednu lekciju za dodatan broj bodova. Posto mi je scena mracnija odlucio sam se za deferred shading.  
Nisam nasao mesta gde se projekat moze dodatno utegnuti, osim sto sam obrisao dosta ponavljajucih linija za postavljanje svetla. Takodje ostaje jos da ocistim projekat od fajlova koji mi nisu potrebni vise, mislim na shadere modela koji se sada crtaju uz pomoc deferred shading-a konkretno: space_objects.glsl i meteor.glsl. Iako su ti fajlovi verizonisani, sacekacu sa brisanjem dok ne kazete da je sve  u redu. 🙂

P.S. Nadam se da nije problem sto je sve u jednom commit-u.

spaske00 commented on engine/src/Framebuffer.cpp:
```diff
@@ -0,0 +1,66 @@
+#include <engine/graphics/OpenGL.hpp>
+#include <engine/platform/PlatformController.hpp>
+
+#include "../libs/assimp/code/AssetLib/MDC/MDCFileData.h"
```
Izgleda da ovo zaglavlje nije potrebno

spaske00 commented on engine/src/Framebuffer.cpp:
```diff
@@ -0,0 +1,66 @@
+
+    Framebuffer::Framebuffer(int width, int height, bool g_buffer) {
+        m_framebuffer = OpenGL::create_framebuffer();
+        if(!g_buffer) {
```
[REAFCTOR] Deluje da nedostaje `else` grana? Da li je u slučaju else grane Framebuffer u validnom stanju i može da se koristi? Ako nije, onda možete ostaviti RG_SHOULD_NOT_REACH_HERE.

perisicMarko replied:
U oba slucaja je framebuffer validan i moze da se koristi. Razlika je u tome sto kada je g_buffer flag true, zauzima se dodatna memorija koja mi nije potrebna za framebuffer koji koristim za postprocesiranje. Zato sam razdvojio da se to zauzima samo kada pravim framebuffer kao g-buffer.
Da li da dodam else granu sa nekim komentarom ili da ostavim ovako obzirom da se bilo kakva radnja izvrsava samo za jednu vrednost flaga?

spaske00 reviewed (changes requested):
Scena izgleda dobro, ostavio sam par komentara.

spaske00 commented:
> Implementirao sam jos jednu lekciju za dodatan broj bodova. Posto mi je scena mracnija odlucio sam se za deferred shading.

Deferred shading je tehnika koja se koristi prilikom računanja osvetljenja za veliki broj svetala na sceni. Scena trenutno ima dva svetla, što nije predviđena upotrebna deffered shadinga. Možete dodati više svetala kao akciju na neko dugme, na primer padanje zvezda/meteora zajedno sa mesecom kada se pritisne E.

perisicMarko commented:
> > Implementirao sam jos jednu lekciju za dodatan broj bodova. Posto mi je scena mracnija odlucio sam se za deferred shading.
> 
> Deferred shading je tehnika koja se koristi prilikom računanja osvetljenja za veliki broj svetala na sceni. Scena trenutno ima dva svetla, što nije predviđena upotrebna deffered shadinga. Možete dodati više svetala kao akciju na neko dugme, na primer padanje zvezda/meteora zajedno sa mesecom kada se pritisne E.

Nisam zavrsio misao zasto sam uzeo deferred shading. "Posto je scena mracnija" nisam siguran koliko bi neka od lekcija sa senkama doprinela izgledu jer nema mnogo svetla. Zbog toga sam uzeo deferred shading, iako ima zanemarljiv uticaj na ovaj broj svetala.

perisicMarko committed:
post review changes

perisicMarko committed:
remove unused functions
update PROJECT-DESCRIPTION-TEMPLATE.md

perisicMarko commented:
Gurnuo sam promene vezane za poslednji review. Da li to sto nisam maksimalno iskoristio tehniku deferred shading znaci da nema osnova za poene? Nisam imao u planu dalji rad na projektu.  Ako je sve ostalo u redu sto se tice toga kakav projekat treba da bude pre nego sto se "preda" sa ovim bih zavrsio.

spaske00 commented on engine/CMakeLists.txt:
```diff
@@ -26,8 +26,8 @@ add_subdirectory(libs/imgui EXCLUDE_FROM_ALL)
-        include/engine/graphics/G_Buffer.cpp
-        include/engine/graphics/G_Buffer.hpp)
+        include/engine/graphics/G_Buffer.hpp
+        src/G_Buffer.cpp)
```
Slobodno možete obrisati ove linije, jer će fajlovi biti pokupljeni sa sa linijom 28. Verovatno ih okruženje dodaje automatski, možete iskljkučiti tu opciju u podešavanjima.

spaske00 commented on PROJECT-DESCRIPTION-TEMPLATE.md:
```diff
@@ -34,7 +34,7 @@ Q -> Lamp on/off switch(works only when moon is turned off)
 
 [x] Bloom with the use of HDR
-[ ] Deferred Shading  
+[X] Deferred Shading  
```
Odštiklirati ovu lekciju s` obzirom da nije implementirana.

spaske00 commented:
> Da li to sto nisam maksimalno iskoristio tehniku deferred shading znaci da nema osnova za poene?   

Boduju se samo lekcije koje su primenljive na datu scenu. Detaljnije informacije u sekciji **Bodovanje** na stranici: https://matf-racunarska-grafika.github.io/projekat/

> Ako je sve ostalo u redu sto se tice toga kakav projekat treba da bude pre nego sto se "preda" sa ovim bih zavrsio.

Samo ostavite ovu granu tako kako jeste, u upustvu na stranici kursa ćete pronaći detaljno upustvo za prijavu i predaju projekta u prvom ispitnom roku.

perisicMarko committed:
changes after final review

perisicMarko committed:
Missing semicolon (FINAL VERSION)
